update for nom version 8

ion0658 opened the pull request:
The nom has been updated to version 8, and the parser api is no longer compatible.

chifflier commented:
Thank you for your contribution, and sorry for the delay.
This is a quite large PR, I am reviewing it and intend to merge as soon as possible.

chifflier commented on Cargo.toml:
```diff
@@ -3,36 +3,37 @@ name = "nom-derive"
-nom = { version = "7.0", default-features = false, features = ["std"] }
-nom-derive-impl = { version="=0.10.0", path="./nom-derive-impl" }
+nom = { version = "8.0", default-features = false, features = ["std"] }
+nom-derive-impl = { version = "=0.10.0", path = "./nom-derive-impl" }
```
This version number should be incremented (as well as in `nom-derive-impl/Cargo.toml`), because crates will not be compatible

chifflier commented on nom-derive-impl/src/config.rs:
```diff
@@ -1,10 +1,11 @@
 
 #[derive(Debug)]
 pub struct Config {
+    #[allow(dead_code)]
```
This is ok for a temporary workaround, but I think we should remove the unused fields (or use them)

chifflier commented on tests/complex.rs:
```diff
@@ -5,17 +5,18 @@ extern crate pretty_assertions;
 struct StructWithComplexParser {
     pub a: u32,
-    #[nom(Parse = "cond(a > 0,be_u64)")]
+    #[nom(Parse = "cond(a > 0,be_u64).parse")]
```
This change could be prevented by moving the call to `.parse` to the generator (see above)

chifflier commented on Cargo.toml:
```diff
@@ -3,36 +3,37 @@ name = "nom-derive"
 repository = "https://github.com/rust-bakery/nom-derive.git"
 readme = "README.md"
-edition = "2018"
+edition = "2024"
```
I have to make more tests, I am not sure moving to edition 2024 is required nor useful. My concern is that I want this crate to stay compatible with older code.
It may be best to keep an old edition but add an MSRV (`rust_version = "1.65"`)

chifflier commented on nom-derive-impl/Cargo.toml:
```diff
@@ -3,33 +3,34 @@ name = "nom-derive-impl"
 
 [dependencies]
-syn = { version="1.0.58", features=["parsing","extra-traits","full"] }
+syn = { version = "1.0.58", features = ["parsing", "extra-traits", "full"] }
```
This may be done in another PR, but `syn` should be updated. I will take care of that

chifflier reviewed (changes requested):
Thank you for your PR!
There are a few things that should be changed before merging. Since it has been some time, it is possible for me to care care of the changes.

ion0658 commented:
- Changed all generators to generate code that calls `.parse`.
- Fixed the CustomError test.
- I incremented the crate version.
- I stopped updating the rust edition.

I'm not sure if I fixed it as intended in the review, since my English is not good enough. If I am wrong, please let me know and I will fix it again.
Thank you!

chifflier commented:
Looks good to me, thanks!
There are a few problems raised by the CI (indentation, MSRV), I will fix them right after merge.

chifflier commented:
Applied, thank you for your contribution
